CMakeLists.txt

```diff
@@ -63,8 +63,21 @@ configure_file("${CMAKE_SOURCE_DIR}/cmake/templates/license.h.in" include/licens
 
 include(EthOptions)
 configure_project(TESTS)
-
+set(LATEST_Z3_VERSION "4.8.12")
 find_package(Z3 4.8.0)
+# NOTE: With EMSCRIPTEN=ON we do not build code from test/ dir and that
+# code is our only reason to require latest Z3.
+if (TESTS AND NOT EMSCRIPTEN)
+  if (${Z3_FOUND})
+    if ( NOT ( "${Z3_VERSION_STRING}" VERSION_EQUAL ${LATEST_Z3_VERSION} ) )
+      message(
+        FATAL_ERROR
+        "SMTChecker tests require Z3 ${LATEST_Z3_VERSION} for all tests to pass.\n\
+Build with -DTESTS=OFF if you want to use an older version. \
```

Comment:
I don't think that's a valid suggestion and not having the latest z3 version, but wanting to run tests is a valid reason for a fatal error (you most definitely should be able to run e.g. ``isoltest --no-smt`` if you don't have the latest Z3 version).

I would suggest to just issue a warning about the Z3 version to be too old for all tests to pass and leave it at that.

Reply:
If we want to be more fancyful, we could set a flag ``OLD_Z3_VERSION`` and have ``soltest`` and ``isoltest`` error out on running themselves without ``--no-smt``.

Reply:
The command line test change is independent of this here anyways, though, so maybe we should split this up into two separate PRs anyways, i.e. reduce this PR to only change the command line tests and then we can separately decide whether any of this should in fact be an error and if so, then how exactly.

Reply:
> I don't think that's a valid suggestion and not having the latest z3 version, but wanting to run tests is a valid thing to do, so it is no reason for a fatal error (you most definitely should be able to run e.g. `isoltest --no-smt` if you don't have the latest Z3 version).
> 
> I would suggest to just issue a warning about the Z3 version to be too old for all tests to pass and leave it at that.

whole point of this PR is to notify the user about the test's requirement to latest z3. I think its better to notify the user as soon as possible, we can change the `FATAL_ERROR` to `WARNING` here and also pass some flag to `soltest` and `isoltest` to error out as you said.

Reply:
> The command line test change is independent of this here anyways, though, so maybe we should split this up into two separate PRs anyways, i.e. reduce this PR to only change the command line tests and then we can separately decide whether any of this should in fact be an error and if so, then how exactly.

good idea, I'll create another PR for that.

Reply:
> not having the latest z3 version, but wanting to run tests is a valid thing to do

Technically yes, but I think that we're better off forcing people to use a specific version of Z3 in this case. In the cases I've seen where we got reports from people getting test failures because of older Z3 they always wanted to run the full suite and were fine with updating Z3 but not with `--no-smt`. They simply did not know that the problems were due to Z3 version. I completely agree with @soroosh-sdi that an error at build stage makes the situation much clearer and easier to resolve on their own for these people.

I suspect that the set of people who care about tests but not Z3 tests and still want to built with Z3 is pretty small. I think that it's more likely that they either do not care about SMTChecker at all (and would be fine with `-DUSE_Z3=OFF`) or do not care about tests at all (and would be fine with `-DTESTS=OFF`).

> good idea, I'll create another PR for that.

Yeah, please create a separate PR for the `--no-smt` flag and rebase this one on top of it. The `--no-smt` change seems to be much less controversial so it'll likely go in without much discussion.

Reply:
I discussed it with @ekpyron on the chat. Here are our conclusions:
1. A build-time check is not enough because you can build with Z3 linked dynamically and then the version used at build time does not matter - you can still get failing tests if you use such a binary with older dynanamic Z3 installed in your system. We need a runtime check in soltest/isoltest that prints an error when `--no-smt` is not used.
    - The build-time check against 4.8.0 is valid because it enforces a version that's still ABI-compatible. That's not the case with the check for latest version, which is rather arbitrary.
2. Turns out there are valid reasons to want older Z3. This changes my view on this check because I thought there's no reason not to always use the latest version. So we should support building with tests enabled and older Z3.
    - But we both agreed that we do not have to make that case the default. We're fine with making it even more stringent. We could require Z3 4.8.12 by default and just add a CMake option that allows the user to override it.
3. We do support `-DTESTS=OFF` but we do not want to encourage users to use it so this part should be removed from the warning.

@soroosh-sdi So here's my suggestion on how to proceed here:
- Create a separate PR for `--no-smt` and we can merge that right away.
- Modify this PR to add a CMake option called `STRICT_Z3_VERSION`.
    - It should be `ON` by default. When it's on, the requirement is 4.8.12, otherwise 4.8.0. And it can be a general requirement, not just for tests (you can drop the `TESTS` check).
        - I think it's not a breaking change but probably should be mentioned in the changelog? @chriseth?
    - In both cases finding an older version should be an error. User should just use `-DUSE_Z3=OFF` if he wants to ignore the installed version. This will require adding back the change you had for Windows CI. This is different than the current behavior (older version is simply ignored right now) but I think it will be better because it's more explicit.
- Let's add the runtime check in `soltest`/`isoltest` in a new PR.

Reply:
Also there's a question if we should still require 4.8.12 when building with dynamic Z3 (i.e. when `USE_Z3_DLOPEN` is enabled). I think that an error in that case is still useful because the user is likely to still run tests in the same system that was used to build the compiler (even if the compiler executable will later be also used on other systems).

Reply:
An interesting question is also what to do if the Z3 version is *too new* (i.e. we haven't yet updated to the latest version or somebody uses a pre-release build from the git tree or somebody builds an old compiler version against more recent Z3). That's just as likely to break the tests as an older version is. So it should probably also require ``-DSTRICT_Z3_VERSION=OFF``.

Reply:
Yes. I agree. This is exactly why I did not name the option `OLD_Z3_VERSION` as you suggested above, but looks like I forgot to mention that the check for 4.8.12 should be an exact comparison. Thanks for pointing this out!

Reply:
Description updated.